cupy/cuda/function.pyx

```diff
@@ -244,3 +261,194 @@ cdef class LinkState:
     cpdef bytes complete(self):
         cubin = driver.linkComplete(self.ptr)
         return cubin
+
+
+cdef class Arg:
+
+    """Base class of internal wrapper of an argument."""
+
+    def __init__(
+            self,
+            object obj,
+            _ArgKind arg_kind,
+            object typ,
+            object dtype,
+            int ndim,
+            bint c_contiguous):
+        self.obj = obj
+        self.arg_kind = arg_kind
+        self.type = typ
+        self.dtype = dtype
+        self.ndim = ndim
+        self.c_contiguous = c_contiguous
+
+    @staticmethod
+    cdef Arg from_obj(object obj):
+        cdef Arg arg
+        # Determine the kind of the argument
+        if isinstance(obj, core.ndarray):
+            arg = NdarrayArg(obj, c_contiguous=obj.flags.c_contiguous)
+            return arg
+        if hasattr(obj, '__cuda_array_interface__'):
+            # __cuda_array_interface__
+            obj = core._convert_object_with_cuda_array_interface(obj)
+            arg = NdarrayArg(obj, c_contiguous=obj.flags.c_contiguous)
+            return arg
+        if _scalar.is_scalar(obj):
+            # scalar
+            arg = ScalarArg(obj)
+            return arg
+        if isinstance(obj, TextureObject):
```

Comment:
Thanks, @asi1024! Could you also include `MemoryPointer` in this category? I have an ongoing PR that would benefit having this. (I could also add this separately later.)
```suggestion
        if isinstance(obj, (TextureObject, MemoryPointer)):
```
(need to import `MemoryPointer` first).

Reply:
I'll fix it after your PR is merged. If this PR is merged before yours, please fix this line in your PR.

Reply:
OK no worries 👍 I am still preparing it...

Reply:
Thanks for your notification!